doc/cursynth.1
.\" Automatically generated by Pod::Man 2.28 (Pod::Simple 3.28)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
.    ds C`
.    ds C'
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.\"
.\" Avoid warning from groff about undefined register 'F'.
.de IX
..
.nr rF 0
.if \n(.g .if rF .nr rF 1
.if (\n(rF:(\n(.g==0)) \{
.    if \nF \{
.        de IX
.        tm Index:\\$1\t\\n%\t"\\$2"
..
.        if !\nF==2 \{
.            nr % 0
.            nr F 2
.        \}
.    \}
.\}
.rr rF
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "I2M20322 1"
.TH I2M20322 1 "2015-05-07" "perl v5.20.2" "GNU Info"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
cursynth.info \- cursynth.info
.SH "GNU Sample"
.IX Header "GNU Sample"
This manual is for cursynth (version 1.6, 7 May 2015).  cursynth is a
music synthesizer.  It's polyphonic and \s-1MIDI\s0 enabled and it runs
graphically within a terminal session.
.PP
* Menu:
.PP
See also \*(L"Invoking cursynth\*(R"
.PP
See also \*(L"Controls\*(R"
.PP
See also \*(L"Synthesis\*(R"
.PP
See also \*(L"Oscillators\*(R"
.PP
See also \*(L"Filter\*(R"
.PP
See also \*(L"Amplifier\*(R"
.PP
See also \*(L"Modulation Matrix\*(R"
.PP
See also \*(L"Effects\*(R"
.PP
See also \*(L"Articulation\*(R"
.PP
See also \*(L"Patches\*(R"
.PP
See also \*(L"\s-1GNU\s0 Free Documentation License\*(R"
.PP
See also \*(L"Index\*(R"
.SS "1 Invoking cursynth"
.IX Subsection "1 Invoking cursynth"
To start cursynth, run:
.PP
$ cursynth
.PP
cursynth will take over the current terminal session and show you all
the synthesis settings as sliders and text selections on the screen.
.PP
Start typing on your keyboard's home row to hear cursynth for the
first time.  Typing F1 or 'H' will show help and information.
.SS "2 Controls"
.IX Subsection "2 Controls"
cursynth is \s-1MIDI\s0 enabled so if you have a working \s-1MIDI\s0 keyboard or
controller, cursynth will respond to note/velocity/etc events and can
learn what parameters to change when a knob or slider is moved through
\&\s-1MIDI\s0 learn.
.PP
* awsedftgyhujkolp;' \- a playable keyboard (no key up events)
* \e'1234567890 \- a slider for the current selected control
* up/down \- previous/next control
* left/right \- decrement/increment control
* F1 (or [shift] + H) \- help/controls
* [shift] + L \- browse/load patches
* [shift] + S \- save patch
* m \- arm \s-1MIDI\s0 learn
* c \- erase \s-1MIDI\s0 learn
.SS "3 Synthesis"
.IX Subsection "3 Synthesis"
cursynth uses a very common form of synthesis called *Subtractive
Synthesis*.  There are three major components in cursynth's synthesis:
.PP
* Oscillators
* Filter
* Amplifier
.PP
Synthesis starts with the two oscillators that create the initial
sound.  The output of the two oscillators are mixed together and fed
into the envelope controlled filter.  The output of the filter is fed
into the envelope controlled amplifier.  Since cursynth is a polyphonic
synthesizer, there is a copy of each component (oscillator, filter,
amplifier) for each voice.  The output of all the voices are combined
into a single output that is fed into the delay line.
.SS "4 Oscillators"
.IX Subsection "4 Oscillators"
The initial sound creation in cursynth happens in the oscillators.
There are two oscillators and you can choose the wave type for each from
the 13 waveforms:
.PP
* sin
* triangle
* square
* down saw
* up saw
* three step
* four step
* eight step
* three pyramid
* five pyramid
* nine pyramid
* white noise
.PP
You can control the second oscillator's transpose relative to the
given note and fine tune the frequency using the tune parameter.
.PP
The two oscillators are hooked into a cross frequency modulation
cycle.  That is, oscillator 1 modifies the frequency of oscillator 2 and
oscillator 2 in turn modifies the frequency of oscillator 1.  Cross
modulation greatly expands the types of sounds you can get from the
oscillators.
.PP
The output of the two oscillators are mixed together using a ratio
definted by the 'oscillator mix' parameter.  A value of 0.5 mixes the
oscillators equally.  A value of 0.0 is only the first oscillator, A
value of 1.0 is only oscillator 2.
.SS "5 Filter"
.IX Subsection "5 Filter"
cursynth has a resonant filter for each voice.  This filter can either
be a low-pass or high-pass filter.
.PP
You can positively or negatively keytrack the filter so that higher
notes will have a higher cutoff and lower notes will have a lower
cutoff.
.PP
The filter also has an \s-1ADSR\s0 envelope that controls the filter cutoff.
It can also be routed to other things using the modulation matrix.
.SS "6 Amplifier"
.IX Subsection "6 Amplifier"
cursynth has an amplifier that is controlled by an \s-1ADSR\s0 amplifier.  You
can also velocity track the notes so the faster you hit a note on a \s-1MIDI\s0
keyboard, the bigger the amplitude of the voice will be.
.SS "7 Modulation Matrix"
.IX Subsection "7 Modulation Matrix"
cursynth has a modulation matrix so you can add custom modulation to
cursynth's routing.  For example, you can set an \s-1LFO\s0 as the source and
the pitch as a destination with a small modulation scale to create a
vibrato.
.PP
There are five fields for five separate modulation settings.
.SS "8 Effects"
.IX Subsection "8 Effects"
cursynth currently has a single delay effect.  The audio of all the
synth voices are combined into one signal that is passed into the delay
effect.
.PP
You can control the feedback amount and delay time.
.SS "9 Articulation"
.IX Subsection "9 Articulation"
cursynth has a number of controls that change how it acts as a
synthesizer.
.PP
* polyphony
* legato
* portamento
* portamento type
* pitch bend range
.PP
polyphony controls the maximum number of active voices at any time.
.PP
legato controls what happens when a note overrides a currently active
note because we had too many active voices.  'off' will cause the note
to retrigger, restarting all oscillators, envelopes and LFOs.  'on' will
not retrigger any oscillators, envelopes or LFOs but will set the
frequency to the new note.
.PP
portamento controls how long it takes to slide the frequency from one
note to another note.  cursynth has a linear (in semitones) slide
measured in seconds-per-octave.
.PP
portamento type controls when the notes slide.  When 'off' notes
never slide their frequency.  When 'on' notes always slide their
frequency.  When 'auto' notes slide only when the note that is being
overridden is still held down on the keyboard, that is will jump to the
new frequency immediately if the overridden note was already released.
.PP
pitch bend range controls how much the pitch wheel affects the pitch
of all the voices, measured in semitones.
.SS "10 Patches"
.IX Subsection "10 Patches"
cursynth is a synthesizer that can make a variety of different sounds
through different settings.  You can save to, and load these settings
from cursynth patches.
.PP
To load, press 'L' (shift + 'l'), then select from the patch list.
You can test the patch by playing on your keyboard.  When you decide on
a patch, press enter to finish loading.
.PP
To save, press 'S' (shift + 's'), then name your patch and hit enter.
The patch will save to \f(CW$HOME\fR/.cursynth/patches/
.PP
All cursynth patches have a .mite extensions and there are two
locations where cursynth looks for them:
.PP
* a system data share directory, (usually
/usr/local/share/cursynth/patches/)
* \f(CW$HOME\fR/.cursynth/patches/
.PP
Patches are also human readable so you can edit them with your text
editor if you wish.
.SS "Appendix A \s-1GNU\s0 Free Documentation License"
.IX Subsection "Appendix A GNU Free Documentation License"
Version 1.2, November 2002
.PP
Copyright (C) 2000,2001,2002 Free Software Foundation, Inc.
59 Temple Place, Suite 330, Boston, \s-1MA  02111\-1307, USA\s0
.PP
Everyone is permitted to copy and distribute verbatim copies
of this license document, but changing it is not allowed.
.PP
0. \s-1PREAMBLE\s0
.PP
The purpose of this License is to make a manual, textbook, or other
functional and useful document \*(L"free\*(R" in the sense of freedom: to
assure everyone the effective freedom to copy and redistribute it,
with or without modifying it, either commercially or
noncommercially.  Secondarily, this License preserves for the
author and publisher a way to get credit for their work, while not
being considered responsible for modifications made by others.
.PP
This License is a kind of \*(L"copyleft\*(R", which means that derivative
works of the document must themselves be free in the same sense.
It complements the \s-1GNU\s0 General Public License, which is a copyleft
license designed for free software.
.PP
We have designed this License in order to use it for manuals for
free software, because free software needs free documentation: a
free program should come with manuals providing the same freedoms
that the software does.  But this License is not limited to
software manuals; it can be used for any textual work, regardless
of subject matter or whether it is published as a printed book.  We
recommend this License principally for works whose purpose is
instruction or reference.
.PP
1. \s-1APPLICABILITY AND DEFINITIONS\s0
.PP
This License applies to any manual or other work, in any medium,
that contains a notice placed by the copyright holder saying it can
be distributed under the terms of this License.  Such a notice
grants a world-wide, royalty-free license, unlimited in duration,
to use that work under the conditions stated herein.  The
\&\*(L"Document\*(R", below, refers to any such manual or work.  Any member
of the public is a licensee, and is addressed as \*(L"you\*(R".  You accept
the license if you copy, modify or distribute the work in a way
requiring permission under copyright law.
.PP
A \*(L"Modified Version\*(R" of the Document means any work containing the
Document or a portion of it, either copied verbatim, or with
modifications and/or translated into another language.
.PP
A \*(L"Secondary Section\*(R" is a named appendix or a front-matter section
of the Document that deals exclusively with the relationship of the
publishers or authors of the Document to the Document's overall
subject (or to related matters) and contains nothing that could
fall directly within that overall subject.  (Thus, if the Document
is in part a textbook of mathematics, a Secondary Section may not
explain any mathematics.)  The relationship could be a matter of
historical connection with the subject or with related matters, or
of legal, commercial, philosophical, ethical or political position
regarding them.
.PP
The \*(L"Invariant Sections\*(R" are certain Secondary Sections whose
titles are designated, as being those of Invariant Sections, in the
notice that says that the Document is released under this License.
If a section does not fit the above definition of Secondary then it
is not allowed to be designated as Invariant.  The Document may
contain zero Invariant Sections.  If the Document does not identify

ChangeLog
2014-03-15  Matt Tytel  <matthewtytel@gmail.com>

	* patches/*: Edited levels of all patches.
	* configure.ac: Bumped version to 1.4
	* src/cursynth_gui.cpp: Updated website to gnu.org/software/cursynth

2014-01-22  gettextize  <bug-gnu-gettext@gnu.org>

	* Makefile.am (ACLOCAL_AMFLAGS): New variable.
	(EXTRA_DIST): Add config.rpath, m4/ChangeLog.
	* configure.ac (AC_CONFIG_FILES): Add po/Makefile.in.
	(AM_GNU_GETTEXT_VERSION): Bump to 0.18.3.

Copyright 2013-2015 Matt Tytel
Copying and distribution of this file, with or without modification, are
promitted provided the copyright notice and this notice are preserved.


src/cursynth.cpp
/* Copyright 2013-2015 Matt Tytel
 *
 * cursynth is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * cursynth is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with cursynth.  If not, see <http://www.gnu.org/licenses/>.
 */

#include "cursynth.h"

#include "cJSON.h"

#include <cstdio>
#include <cstdlib>
#include <dirent.h>
#include <fstream>
#include <iostream>
#include <ncurses.h>
#include <sstream>
#include <string>
#include <sys/stat.h>

#define KEYBOARD "awsedftgyhujkolp;'"
#define SLIDER "`1234567890"
#define EXTENSION ".mite"
#define CONFIG_DIR ".cursynth/"
#define USER_PATCHES_DIR "patches/"
#define CONFIG_FILE ".cursynth_conf"
#define NUM_CHANNELS 2
#define MOD_WHEEL_ID 1
#define PITCH_BEND_PORT 224
#define SUSTAIN_PORT 176
#define SUSTAIN_ID 64

namespace {

  // Receive MIDI data and send it to the synth.
  void midiCallback(double delta_time, std::vector<unsigned char>* message,
                    void* user_data) {
    UNUSED(delta_time);
    mopo::Cursynth* cursynth = static_cast<mopo::Cursynth*>(user_data);
    cursynth->processMidi(message);
  }

  // Receive audio buffers and send them to the synth.
  int audioCallback(void *out_buffer, void *in_buffer,
                    unsigned int n_frames, double stream_time,
                    RtAudioStreamStatus status, void *user_data) {
    UNUSED(in_buffer);
    UNUSED(stream_time);
    if (status)
      std::cout << "Stream underflow detected!" << std::endl;

    mopo::Cursynth* cursynth = static_cast<mopo::Cursynth*>(user_data);
    cursynth->processAudio((mopo::mopo_float*)out_buffer, n_frames);
    return 0;
  }

  // Return the directory containing the configuration file and patches.
  std::string getConfigPath() {
    std::stringstream config_path;
    config_path << getenv("HOME") << "/" << CONFIG_DIR;
    return config_path.str();
  }

  // Return the location of the configuration file.
  std::string getConfigFile() {
    std::stringstream config_path;
    config_path << getConfigPath() << CONFIG_FILE;
    return config_path.str();
  }

  // Return the location of the user patches.
  std::string getUserPatchesPath() {
    std::stringstream patches_path;
    patches_path << getConfigPath() << USER_PATCHES_DIR;
    return patches_path.str();
  }

  // Check if the directory _path_ exists, if not, create it.
  void confirmPathExists(std::string path) {
    if (opendir(path.c_str()) == NULL)
      mkdir(path.c_str(), 0755);
  }

  // Returns all files in directory _dir_ with extenstion _ext_.
  std::vector<std::string> getAllFiles(std::string dir, std::string ext) {
    std::vector<std::string> file_names;
    DIR *directory = NULL;
    struct dirent *ent = NULL;

    if ((directory = opendir(dir.c_str())) != NULL) {
      while ((ent = readdir(directory)) != NULL) {
        std::string name = ent->d_name;
        if (name.find(ext) != std::string::npos)
          file_names.push_back(name);
      }
      closedir(directory);
    }

    return file_names;
  }
} // namespace

namespace mopo {
  Cursynth::Cursynth() : state_(STANDARD), patch_load_index_(0) {
    pthread_mutex_init(&mutex_, 0);
  }

  void Cursynth::start(unsigned sample_rate, unsigned buffer_size) {
    // Setup all callbacks.
    setupAudio(sample_rate, buffer_size);
    setupMidi();
    setupGui();
    loadConfiguration();

    // Wait for computer keyboard input.
    while(textInput(getch()))
      ;

    stop();
  }

  void Cursynth::loadConfiguration() {
    // Try to open the configuration file.
    std::ifstream config_file;
    config_file.open(getConfigFile().c_str());
    if (!config_file.is_open())
      return;

    // Read file contents into buffer.
    config_file.seekg(0, std::ios::end);
    int length = config_file.tellg();
    config_file.seekg(0, std::ios::beg);
    char file_contents[length];
    config_file.read(file_contents, length);

    // Parse the JSON configuration file.
    cJSON* root = cJSON_Parse(file_contents);

    // For all controls, try to load the MIDI learn assignment.
    std::string current = gui_.getCurrentControl();
    std::string name = current;
    do {
      cJSON* value = cJSON_GetObjectItem(root, name.c_str());
      if (value)
        midi_learn_[value->valueint] = name;

      name = gui_.getNextControl();
    } while(name != current);

    // Delete the parsing data.
    cJSON_Delete(root);
  }

  void Cursynth::saveConfiguration() {
    confirmPathExists(getConfigPath());

    // Store all the MIDI learn data into JSON.
    cJSON* root = cJSON_CreateObject();
    std::map<int, std::string>::iterator iter = midi_learn_.begin();
    for (; iter != midi_learn_.end(); ++iter) {
      cJSON* midi = cJSON_CreateNumber(iter->first);
      cJSON_AddItemToObject(root, iter->second.c_str(), midi);
    }

    // Write the configuration JSON to the configuration file.
    char* json = cJSON_Print(root);
    std::ofstream save_file;
    save_file.open(getConfigFile().c_str());
    MOPO_ASSERT(save_file.is_open());
    save_file << json;
    save_file.close();

    free(json);
    cJSON_Delete(root);
  }

  bool Cursynth::textInput(int key) {
    if (state_ == PATCH_LOADING) {
      int num_patches = patches_.size();
      switch(key) {
        case '\n':
          // Finish loading patches.
          state_ = STANDARD;
          gui_.clearPatches();
          return true;
        case KEY_UP:
          // Go to previous patch.
          patch_load_index_ = CLAMP(patch_load_index_ - 1, 0, num_patches - 1);
          gui_.drawPatchLoading(patches_, patch_load_index_);
          loadFromFile(patches_[patch_load_index_]);
          return true;
        case KEY_DOWN:
          // Go to next patch.
          patch_load_index_ = CLAMP(patch_load_index_ + 1, 0, num_patches - 1);
          gui_.drawPatchLoading(patches_, patch_load_index_);
          loadFromFile(patches_[patch_load_index_]);
          return true;
      }
    }

    std::string current_control = gui_.getCurrentControl();
    Control* control = controls_.at(current_control);
    bool should_redraw_control = false;
    lock();
    switch(key) {
      case 'H':
      case KEY_F(1):
        startHelp();
        break;
      case 'S':
        startSave();
        break;
      case 'L':
        startLoad();
        break;
      case 'M':
      case 'm':
        if (state_ != MIDI_LEARN)
          state_ = MIDI_LEARN;
        else
          state_ = STANDARD;
        should_redraw_control = true;
        break;
      case 'C':
      case 'c':
        eraseMidiLearn(control);
        state_ = STANDARD;
        should_redraw_control = true;
        break;
      case KEY_UP:
        current_control = gui_.getPrevControl();
        state_ = STANDARD;
        gui_.drawControl(control, false);
        should_redraw_control = true;
        break;
      case KEY_DOWN:
        current_control = gui_.getNextControl();
        state_ = STANDARD;
        gui_.drawControl(control, false);
        should_redraw_control = true;
        break;
      case KEY_RIGHT:
        control->increment();
        should_redraw_control = true;
        break;
      case KEY_LEFT:
        control->decrement();
        should_redraw_control = true;
        break;
      case KEY_RESIZE:
        refreshGui();
        break;
      default:
        // Check if they pressed the slider keys and change the current value.
        size_t slider_size = strlen(SLIDER) - 1;
        for (size_t i = 0; i <= slider_size; ++i) {
          if (SLIDER[i] == key) {
            control->setPercentage((1.0 * i) / slider_size);
            should_redraw_control = true;
            break;
          }
        }

        // Check if they pressed note keys and play the corresponding note.
        for (size_t i = 0; i < strlen(KEYBOARD); ++i) {
          if (KEYBOARD[i] == key) {
            synth_.noteOn(48 + i);
            break;
          }
        }
    }

    if (should_redraw_control) {
      control = controls_.at(current_control);
      gui_.drawControl(control, true);
      gui_.drawControlStatus(control, state_ == MIDI_LEARN);
    }

    unlock();
    return true;
  }

  unsigned int Cursynth::chooseSampleRate(const RtAudio::DeviceInfo& device,
                                          unsigned preferred_sample_rate) {
    for (int i = 0; i < device.sampleRates.size(); ++i) {
      if (device.sampleRates[i] == preferred_sample_rate)
        return preferred_sample_rate;
    }

    return device.sampleRates[0];
  }

  void Cursynth::setupAudio(unsigned sample_rate, unsigned buffer_size) {
    // Make sure we have a device to make sound with.
    if (dac_.getDeviceCount() < 1) {
      std::cout << "No audio devices found.\n";
      exit(0);
    }

    // Setup audio preferences.
    RtAudio::StreamParameters parameters;
    parameters.deviceId = dac_.getDefaultOutputDevice();
    parameters.nChannels = NUM_CHANNELS;
    parameters.firstChannel = 0;
    RtAudio::DeviceInfo device_info = dac_.getDeviceInfo(parameters.deviceId);

    unsigned actual_sample_rate = chooseSampleRate(device_info, sample_rate);
    synth_.setSampleRate(actual_sample_rate);
    buffer_size = CLAMP(buffer_size, 0, mopo::MAX_BUFFER_SIZE);

    // Start the audio callbacks.
    try {
      dac_.openStream(&parameters, NULL, RTAUDIO_FLOAT64, actual_sample_rate,
                      &buffer_size, &audioCallback, (void*)this);
      dac_.startStream();
    }
    catch (RtError& error) {
      error.printMessage();
      exit(0);
    }

    synth_.setBufferSize(buffer_size);
  }

  void Cursynth::setupGui() {
    gui_.start();

    // Add the controls to the GUI for viewing.
    controls_ = synth_.getControls();
    gui_.addControls(controls_);

    // Make sure we are drawing he current control.
    Control* control = controls_.at(gui_.getCurrentControl());
    gui_.drawControl(control, true);
    gui_.drawControlStatus(control, false);
  }

  void Cursynth::refreshGui() {
    gui_.redrawBase();
    controls_ = synth_.getControls();
    gui_.addControls(controls_);

    // Make sure we are drawing the current control.
    Control* control = controls_.at(gui_.getCurrentControl());
    gui_.drawControl(control, true);
    gui_.drawControlStatus(control, false);
  }

  void Cursynth::processAudio(mopo_float *out_buffer, unsigned int n_frames) {
    // Run the synth.
    lock();
    synth_.process();
    unlock();

    // Copy the synth output to the output buffer.
    const mopo_float* buffer = synth_.output()->buffer;
    for (size_t i = 0; i < n_frames; ++i) {
      for (int c = 0; c < NUM_CHANNELS; ++c)
        out_buffer[NUM_CHANNELS * i + c] = buffer[i];
    }
  }

  void Cursynth::eraseMidiLearn(Control* control) {
    if (control->midi_learn()) {
      midi_learn_.erase(control->midi_learn());
      control->midi_learn(0);
    }
  }

  void Cursynth::setupMidi() {
    RtMidiIn* midi_in = new RtMidiIn();
    if (midi_in->getPortCount() <= 0) {
      std::cout << "No midi devices found.\n";
    }

    // Setup MIDI callbacks for every MIDI device.
    // TODO: Have a menu for only enabling some MIDI devices.
    for (unsigned int i = 0; i < midi_in->getPortCount(); ++i) {
      RtMidiIn* device = new RtMidiIn();
      device->openPort(i);
      device->setCallback(&midiCallback, (void*)this);
      midi_ins_.push_back(device);
    }

    delete midi_in;
  }

  void Cursynth::processMidi(std::vector<unsigned char>* message) {
    if (message->size() < 3)
      return;

    lock();
    int midi_port = message->at(0);
    int midi_id = message->at(1);
    int midi_val = message->at(2);
    std::string selected_control_name = gui_.getCurrentControl();
    Control* selected_control = controls_.at(selected_control_name);
    if (midi_port >= 144 && midi_port < 160) {
      // A MIDI keyboard key was pressed. Play a note.
      int midi_note = midi_id;
      int midi_velocity = midi_val;

      if (midi_velocity)
        synth_.noteOn(midi_note, (1.0 * midi_velocity) / MIDI_SIZE);
      else
        synth_.noteOff(midi_note);
    }
    else if (midi_port >= 128 && midi_port < 144) {
      // A MIDI keyboard key was released. Release that note.
      int midi_note = midi_id;
      synth_.noteOff(midi_note);
    }
    else if (midi_port == PITCH_BEND_PORT)
      synth_.setPitchWheel((2.0 * midi_val) / (MIDI_SIZE - 1) - 1);
    else if (midi_port == SUSTAIN_PORT && midi_id == SUSTAIN_ID) {
      if (midi_val)
        synth_.sustainOn();
      else
        synth_.sustainOff();
    }
    else if (midi_port < 254) {
      // Must have gotten MIDI from some knob or other control.
      if (state_ == MIDI_LEARN && midi_port < 254) {
        // MIDI learn is armed so map this MIDI signal to the current control.
        eraseMidiLearn(selected_control);

        midi_learn_[midi_id] = selected_control_name;
        selected_control->midi_learn(midi_id);
        state_ = STANDARD;
        gui_.drawControlStatus(selected_control, false);
        saveConfiguration();
      }
      else if (midi_learn_.find(midi_id) != midi_learn_.end()) {
        // MIDI learn is enabled for this control. Change the paired control.
        Control* midi_control = controls_.at(midi_learn_[midi_id]);
        midi_control->setMidi(midi_val);
        gui_.drawControl(midi_control, selected_control == midi_control);
        gui_.drawControlStatus(midi_control, false);
      }

      if (midi_id == MOD_WHEEL_ID)
        synth_.setModWheel(midi_val);
    }
    unlock();
  }

  void Cursynth::stop() {
    pthread_mutex_destroy(&mutex_);
    gui_.stop();
    try {
      dac_.stopStream();
    }
    catch (RtError& error) {
      error.printMessage();
    }

    if (dac_.isStreamOpen())
      dac_.closeStream();
  }

  void Cursynth::startHelp() {
    gui_.drawHelp();
    getch();

    gui_.drawMain();
    std::string current = gui_.getCurrentControl();
    std::string name = gui_.getNextControl();
    while (name != current) {
      gui_.drawControl(controls_.at(name), false);
      name = gui_.getNextControl();
    }
    gui_.drawControl(controls_.at(current), true);
  }

  void Cursynth::startSave() {
    gui_.clearPatches();
    curs_set(1);
    std::stringstream save_stream;
    gui_.drawPatchSaving(save_stream.str());
    int key = 0;
    while(1) {
      key = getch();
      std::string current = save_stream.str();
      switch(key) {
        case KEY_DC:
        case KEY_BACKSPACE:
        case 127:
          // Pressed delete, so remove the last character from the file name.
          save_stream.str(current.substr(0, current.length() - 1));
          save_stream.seekp(save_stream.str().length());
          gui_.drawPatchSaving(save_stream.str());
          break;
        case '\n':
          // Pressed enter, so save to the current file name.
          // Then go back to the normal input state.
          if (save_stream.str().length() > 0)
            saveToFile(save_stream.str() + EXTENSION);
          state_ = STANDARD;
          gui_.clearPatches();
          curs_set(0);
          return;
        default:
          // If it's a printable character, append it to the file name.
          if (isprint(key))
            save_stream << static_cast<char>(key);
          gui_.drawPatchSaving(save_stream.str());
          break;
      }
    }
  }

  void Cursynth::startLoad() {
    // Store all patche names from system and user patch directories.
    patches_ = getAllFiles(PATCHES_DIRECTORY, EXTENSION);
    std::vector<std::string> user_patches =
        getAllFiles(getUserPatchesPath(), EXTENSION);
    patches_.insert(patches_.end(), user_patches.begin(), user_patches.end());

    if (patches_.size() == 0)
      return;

    // Start patch loading by loading last patch browsed.
    state_ = PATCH_LOADING;
    patch_load_index_ = std::min<int>(patch_load_index_, patches_.size() - 1);
    loadFromFile(patches_[patch_load_index_]);
  }

  void Cursynth::saveToFile(const std::string& file_name) {
    confirmPathExists(getConfigPath());
    confirmPathExists(getUserPatchesPath());
    std::ofstream save_file;
    std::string path = getUserPatchesPath();
    path = path + "/" + file_name;
    save_file.open(path.c_str());
    save_file << writeStateToString();
    save_file.close();
  }

  void Cursynth::loadFromFile(const std::string& file_name) {
    std::ifstream load_file;

    // First try to load the patch from user patches.
    std::string path = getUserPatchesPath();
    path += "/";
    path += file_name;
    load_file.open(path.c_str());

    // If we can't find the patch try default system patches.
    if (!load_file.is_open()) {
      path = PATCHES_DIRECTORY;
      path += "/";
      path += file_name;
      load_file.open(path.c_str());
    }

    if (!load_file.is_open())
      return;

    // Read file into buffer and load it.
    load_file.seekg(0, std::ios::end);
    int length = load_file.tellg();
    load_file.seekg(0, std::ios::beg);
    char file_contents[length];
    load_file.read(file_contents, length);
    readStateFromString(file_contents);
    load_file.close();

    // Draw the patch loading happen.
    gui_.drawPatchLoading(patches_, patch_load_index_);
  }

  std::string Cursynth::writeStateToString() {
    // Write all controls to JSON.
    cJSON* root = cJSON_CreateObject();
    control_map::iterator iter = controls_.begin();
    for (; iter != controls_.end(); ++iter) {
      cJSON* value = cJSON_CreateNumber(iter->second->value()->value());
      cJSON_AddItemToObject(root, iter->first.c_str(), value);
    }

    // Write the JSON to a string and free the JSON.
    char* json = cJSON_Print(root);
    std::string output = json;
    free(json);
    cJSON_Delete(root);

    return output;
  }

  void Cursynth::readStateFromString(const std::string& state) {
    // Parse state into JSON and read all controls.
    cJSON* root = cJSON_Parse(state.c_str());
    control_map::iterator iter = controls_.begin();
    for (; iter != controls_.end(); ++iter) {
      cJSON* value = cJSON_GetObjectItem(root, iter->first.c_str());

      if (value) {
        Control* control = iter->second;
        control->set(value->valuedouble);
        gui_.drawControl(control, false);
      }
    }

    // Setup current control.
    Control* current_control = controls_.at(gui_.getCurrentControl());
    gui_.drawControl(current_control, true);
    gui_.drawControlStatus(current_control, false);

    cJSON_Delete(root);
  }
} // namespace mopo
